frontend/src/metabase/nav/containers/MainNavbar/MainNavbarContainer/AddDataModal/Panels/CSVUpload.tsx

```diff
@@ -0,0 +1,299 @@
+            file: null,
+            error: t`Sorry, this file is too large`,
+          });
+          break;
+        default:
+          setUploadState({ file: null, error: "An error has occurred" });
+          break;
+      }
+    }
+  }, []);
+
+  const onDrop = useCallback(
+    (acceptedFiles: File[], fileRejections: FileRejection[]) => {
+      if (acceptedFiles.length === 1) {
+        setUploadState({ file: acceptedFiles[0], error: null });
+      }
+
+      handleFileRejections(fileRejections);
+    },
+    [handleFileRejections],
+  );
+
+  const dropZoneConfig = useMemo(
+    () => ({
+      onDrop,
+      maxFiles: 1,
+      maxSize: MAX_UPLOAD_SIZE,
+      noClick: true,
+      noDragEventsBubbling: true,
+      accept: { "text/csv": [".csv"], "text/tab-separated-values": [".tsv"] },
+    }),
+    [onDrop],
+  );
+
+  const { getRootProps, isDragActive } = useDropzone(dropZoneConfig);
+
+  const dropzoneProps = getComposedDragProps(getRootProps());
+
+  const handleFileInput = useCallback(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      const file = event.target.files?.[0];
+
+      if (!file) {
+        return;
+      }
+
+      if (file.size > MAX_UPLOAD_SIZE) {
+        setUploadState({ file: null, error: t`Sorry, this file is too large` });
+        return;
+      }
+
+      setUploadState({ file, error: null });
+
+      // reset the input so that the same file can be uploaded again
+      if (uploadInputRef.current) {
+        uploadInputRef.current.value = "";
+      }
+    },
+    [],
+  );
+
+  const handleFileUpload = useCallback(
+    (uploadedFile: File | null) => {
+      if (!uploadedFile) {
+        return;
+      }
+
+      dispatch(
+        uploadFile({
+          uploadMode: UploadMode.create,
+          collectionId: uploadCollectionId,
+          file: uploadedFile,
+        }),
+      );
+
+      setUploadState({ file: null, error: null });
+      onCloseAddDataModal();
+    },
+    [dispatch, onCloseAddDataModal, uploadCollectionId],
+  );
+
+  const primaryText = useMemo(() => {
+    if (uploadState.file) {
+      return uploadState.file.name;
+    }
+
+    if (uploadState.error) {
+      return uploadState.error;
+    }
+
+    return t`Drag and drop a file here`;
+  }, [uploadState]);
+
+  return (
+    <>
+      <Stack align="stretch" justify="space-between" h="100%" w="100%">
+        <Center
+          ta="center"
+          {...dropzoneProps}
+          className={cx(S.dropZone, isDragActive && S.isActive)}
+          data-testid="add-data-modal-csv-dropzone"
+        >
+          <Stack gap="sm" align="center">
+            {uploadState.error ? (
+              <Box component={IconCSVWarning} h={50} />
+            ) : (
+              <Box
+                c={uploadState.file ? "brand" : "text-secondary-inverse"}
+                component={IconCSV}
+                h={50}
+              />
+            )}
+
+            <div>
+              <Text fw={700}>{primaryText}</Text>
```

Comment:
I think we should stay away from numeric weights
```suggestion
              <Text fw="bold">{primaryText}</Text>
```